File: Raven.Client.Lightweight/IDocumentStore.cs
//-----------------------------------------------------------------------
// <copyright file="IDocumentStore.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
#if !SILVERLIGHT
using System.Collections.Specialized;
#endif
using System.Collections.Generic;
using System.Net;
using Raven.Client.Connection;
using Raven.Client.Connection.Profiling;
using Raven.Client.Document;
#if SILVERLIGHT
using Raven.Client.Silverlight.Connection;
#else
using Raven.Client.Indexes;
#endif
#if !NET35
using Raven.Client.Connection.Async;
#endif

namespace Raven.Client
{
	/// <summary>
	/// Interface for managing access to RavenDB and open sessions.
	/// </summary>
	public interface IDocumentStore : IDisposalNotification
	{
		/// <summary>
		/// Setup the context for aggressive caching.
		/// </summary>
		/// <param name="cahceDuration">Specify the aggressive cache duration</param>
		/// <remarks>
		/// Aggressive caching means that we will not check the server to see whatever the response
		/// we provide is current or not, but will serve the information directly from the local cache
		/// without touching the server.
		/// </remarks>
		IDisposable AggressivelyCacheFor(TimeSpan cahceDuration);

		/// <summary>
		/// Setup the context for no aggressive caching
		/// </summary>
		/// <remarks>
		/// This is mainly useful for internal use inside RavenDB, when we are executing
		/// queries that has been marked with WaitForNonStaleResults, we temporarily disable
		/// aggressive caching.
		/// </remarks>
		IDisposable DisableAggressiveCaching();

		/// <summary>
		/// Gets the shared operations headers.
		/// </summary>
		/// <value>The shared operations headers.</value>
#if !SILVERLIGHT
		NameValueCollection SharedOperationsHeaders { get; }
#else
		IDictionary<string,string> SharedOperationsHeaders { get; }
#endif

		/// <summary>
		/// Get the <see cref="HttpJsonRequestFactory"/> for this store
		/// </summary>
		HttpJsonRequestFactory JsonRequestFactory { get; }

		/// <summary>
		/// Gets or sets the identifier for this store.
		/// </summary>
		/// <value>The identifier.</value>
		string Identifier { get; set; }

		/// <summary>
		/// Initializes this instance.
		/// </summary>
		/// <returns></returns>
		IDocumentStore Initialize();


#if !NET35
		/// <summary>
		/// Gets the async database commands.
		/// </summary>
		/// <value>The async database commands.</value>
		IAsyncDatabaseCommands AsyncDatabaseCommands { get; }

		/// <summary>
		/// Opens the async session.
		/// </summary>
		/// <returns></returns>
		IAsyncDocumentSession OpenAsyncSession();

		/// <summary>
		/// Opens the async session.
		/// </summary>
		/// <returns></returns>
		IAsyncDocumentSession OpenAsyncSession(string database);
#endif

#if !SILVERLIGHT
		/// <summary>
		/// Opens the session.
		/// </summary>
		/// <returns></returns>
		IDocumentSession OpenSession();

		/// <summary>
		/// Opens the session for a particular database
		/// </summary>
		IDocumentSession OpenSession(string database);

		/// <summary>
		/// Opens the session with the specified options.
		/// </summary>
		IDocumentSession OpenSession(OpenSessionOptions sessionOptions);

		/// <summary>
		/// Gets the database commands.
		/// </summary>
		/// <value>The database commands.</value>
		IDatabaseCommands DatabaseCommands { get; }

		
		/// <summary>
		/// Executes the index creation.
		/// </summary>
		void ExecuteIndex(AbstractIndexCreationTask indexCreationTask);
#endif

		/// <summary>
		/// Gets the conventions.
		/// </summary>
		/// <value>The conventions.</value>
		DocumentConvention Conventions { get; }

		/// <summary>
		/// Gets the URL.
		/// </summary>
		string Url { get; }

		///<summary>
		/// Gets the etag of the last document written by any session belonging to this 
		/// document store
		///</summary>
		Guid? GetLastWrittenEtag();
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/IDocumentStore.java
package net.ravendb.client;

import java.util.Map;

import net.ravendb.abstractions.data.BulkInsertOptions;
import net.ravendb.abstractions.data.Etag;
import net.ravendb.client.changes.IDatabaseChanges;
import net.ravendb.client.connection.implementation.HttpJsonRequestFactory;
import net.ravendb.client.document.BulkInsertOperation;
import net.ravendb.client.document.DocumentConvention;
import net.ravendb.client.document.OpenSessionOptions;
import net.ravendb.client.indexes.AbstractIndexCreationTask;
import net.ravendb.client.indexes.AbstractTransformerCreationTask;


/**
 * Interface for managing access to RavenDB and open sessions.
 */
public interface IDocumentStore extends IDisposalNotification {

  /**
   * Subscribe to change notifications from the server
   * @return
   */
  public IDatabaseChanges changes();

  /**
   * Subscribe to change notifications from the server
   * @param database
   * @return
   */
  public IDatabaseChanges changes(String database);

  /**
   * Setup the context for aggressive caching.
   *
   * Aggressive caching means that we will not check the server to see whatever the response
   * we provide is current or not, but will serve the information directly from the local cache
   * without touching the server.
   * @param cacheDurationInMilis
   * @return
   */
  public AutoCloseable aggressivelyCacheFor(long cacheDurationInMilis);

  /**
   * Setup the context for aggressive caching.
   *
   * Aggressive caching means that we will not check the server to see whatever the response
   * we provide is current or not, but will serve the information directly from the local cache
   * without touching the server.
   * @return
   */
  public AutoCloseable aggressivelyCache();

  /**
   * Setup the context for no aggressive caching
   *
   * This is mainly useful for internal use inside RavenDB, when we are executing
   * queries that has been marked with WaitForNonStaleResults, we temporarily disable
   * aggressive caching.
   * @return
   */
  public AutoCloseable disableAggressiveCaching();

  /**
   * Setup the WebRequest timeout for the session
   * @param timeout Specify the timeout duration
   * @return Sets the timeout for the JsonRequest.  Scoped to the Current Thread.
   */
  public AutoCloseable setRequestsTimeoutFor(long timeout);

  /**
   * Gets the shared operations headers.
   * @return
   */
  public Map<String, String> getSharedOperationsHeaders();


  /**
   * Get the {@link HttpJsonRequestFactory} for this store
   * @return
   */
  public HttpJsonRequestFactory getJsonRequestFactory();

  /**
   * Sets the identifier
   * @param identifier
   */
  public void setIdentifier(String identifier);

  /**
   * Gets the identifier
   * @return
   */
  public String getIdentifier();

  /**
   * Initializes this instance.
   * @return
   */
  public IDocumentStore initialize();

  /**
   * Opens the session.
   * @return
   */
  public IDocumentSession openSession();

  /**
   * Opens the session for a particular database
   * @param database
   * @return
   */
  public IDocumentSession openSession(String database);

  /**
   * Opens the session with the specified options.
   * @param sessionOptions
   * @return
   */
  public IDocumentSession openSession(OpenSessionOptions sessionOptions);

  /**
   * Gets the database commands.
   * @return
   */
  public net.ravendb.client.connection.IDatabaseCommands getDatabaseCommands();

  /**
   * Executes the index creation.
   * @param indexCreationTask
   */
  public void executeIndex(AbstractIndexCreationTask indexCreationTask);

  /**
   * executes the transformer creation
   * @param transformerCreationTask
   */
  public void executeTransformer(AbstractTransformerCreationTask transformerCreationTask);

  /**
   * Gets the conventions.
   * @return
   */
  public DocumentConvention getConventions();

  /**
   * Gets the URL.
   * @return
   */
  public String getUrl();

  /**
   * Gets the etag of the last document written by any session belonging to this
   * document store
   * @return
   */
  public Etag getLastWrittenEtag();

  /**
   * Performs bulk insert
   * @return
   */
  public BulkInsertOperation bulkInsert();

  /**
   * Performs bulk insert
   * @param database
   * @return
   */
  public BulkInsertOperation bulkInsert(String database);

  /**
   * Performs bulk insert
   * @param database
   * @param options
   * @return
   */
  public BulkInsertOperation bulkInsert(String database, BulkInsertOptions options);

}
